model/variableSheet.cc

```diff
@@ -32,184 +31,48 @@ using ecolab::cairo::CairoSave;
-                              { // draw vertical grid line
-                                cairo::CairoSave cs(cairo);
-                                cairo_set_source_rgba(cairo,0,0,0,0.5);
-                                cairo_move_to(cairo,x-2.5,y0);
-                                cairo_line_to(cairo,x-2.5,y0+lh+1.1*rowHeight);
-                                cairo_stroke(cairo);
-                              }
-                              colWidth=std::max(colWidth, 5+pango.width());
-                              for (size_t j=0; j<dims[0]; ++j)
-                                {
-                                  y+=rowHeight;
-                                  if (y>m_height) break;
-                                  cairo_move_to(cairo,x,y);
-                                  auto v=value->atHCIndex(j+i*dims[0]);
-                                  if (!std::isnan(v))
-                                    {
-                                      pango.setText(str(v));
-                                      pango.show();
-                                    }
-                                  colWidth=std::max(colWidth, pango.width());
-                                }
-                              x+=colWidth;
-                              if (x>m_width) break;
-                            }      
-                        }
-                      h_prev=h;  
-                      w=0;h=0;      
-                      cairo_get_current_point (cairo,&w,&h);   
-                      if (h<h_prev) h+=h_prev;                                                                         
-                      cout << " " << w << " " << h << " " << y0 << " " << h_prev << endl;
-                      // draw grid
-                      {
-                        cairo::CairoSave cs(cairo);
-                        cairo_set_source_rgba(cairo,0,0,0,0.2);
-                        for (y=y0+rowHeight; y<h+rowHeight; y+=2*rowHeight)
-                          {
-                            cairo_rectangle(cairo,0.0,y,w+colWidth,rowHeight);
-                            cairo_fill(cairo);
-                          }
-                      }
-                
-                    }
-                  cairo::CairoSave cs(cairo);    
-                  cairo_rectangle(cairo,x0,y0,w+colWidth,y-y0);    
-                  cairo_stroke(cairo);                          
-                  cairo_clip(cairo);
-                  y0=h+1.1*rowHeight;   
-                }      
-            }
-        }
-    }
-  catch (...) {throw;/* exception most likely invalid variable value */}
-}        
+  void VariableSheet::populateItemVector() {
+    itemVector.clear();	
+    minsky().canvas.model->recursiveDo(&GroupItems::items,
+                                       [&](Items&, Items::iterator i) {
+                                         if ((*i)->variableCast()->type()!=VariableType::parameter)		                                 
```

Comment:
Two approaches you can use to share between variablesheet and  parametersheet
1. Add virtual selector function to the common base class that returns returns v && v->type()==parameter / v && v->type()!=parameter respectively in parameterSheet and VariableSheet. Then call that with the argument (*i)->variableCast() in this lambda (which can now be moved to the common base class. Note the presence of the guard v - as it currently stands, this code will do a null dereference whenever the item is not a variable.

Reply:
Thank you, that is a very useful suggestion.